c/shared/source/uforead/uforead.c

```diff
@@ -1813,10 +1951,72 @@ static bool parseFontInfoFDArray(ufoCtx h, xmlNodePtr cur) {
     return true;
 }
 
+/* set each GLIF_REC's iFD based on the FDArraySelect groups*/
+static bool parseFDArraySelectGroup(ufoCtx h, char* FDArraySelectKeyName, xmlNodePtr cur) {
+    GLIF_Rec *g;
+
+    if (!xmlStrEqual((cur)->name, (const xmlChar *) "array"))
+        return false;
+
+    char* FDIndex = memNew(h, sizeof(char) * 2);
+    strncpy(FDIndex, FDArraySelectKeyName + 14, 2);
+    if (FDIndex[1] == '.')
+        FDIndex[1] = '\n';
+    int FDIndexInt = atoi(FDIndex);
+
+    /* check if FDArray is defined */
+    if (h->top.FDArray.cnt - 1 <= FDIndexInt)
+        fatal(h, ufoErrParseFail, "Number of FDArrays defined in lib.plist not equal to number of FDSelect Groups in groups.plist.");
+
+    cur = cur->xmlChildrenNode;
+    while (cur != NULL) {
+        char* glyphName = parseXMLKeyValue(h, cur);
+        g = findGLIFRecByName(h, glyphName);
+        if (g != NULL) {
+            if (FDIndexInt >= h->top.FDArray.cnt - 1) {  /* one cnt reserved for default FDArray at this point */
+//                message(h, "Glyph %s is assigned to font dict index %d which is not defined. Will be assigned to default font dict at index 0 instead.\n", glyphName, FDIndexInt);
+                FDIndexInt  = 0;
+            }
+            g->iFD = FDIndexInt;
+        }
+        cur = cur->next;
+    }
+    memFree(h, FDIndex);
+    return true;
+}
+
+static bool parseCIDMap(ufoCtx h, xmlNodePtr cur) {
+    h->parseState.CIDMap = true;
+    GLIF_Rec *g;
+
+    if (!xmlStrEqual((cur)->name, (const xmlChar *) "dict"))
+        return false;
+
+    cur = cur->xmlChildrenNode;
+    while (cur != NULL) {
+        char* keyName = parseXMLKeyName(h, cur);
+        if (keyName != NULL) {
+            g = addNewGLIFRec(h, keyName, NULL);
+            if (g == NULL)
+                continue;
+            cur = cur->next;
+            if (cur == NULL)
+                continue;
+            char* cid = parseXMLKeyValue(h, cur);
+            if (cid != NULL)
+                g->cid = atoi(cid);
```

Comment:
This is probably OK as garbage will result in a 0 value that will be detected later, but it's better to sanity-check whether a 0 value makes sense. (Can we tell at this point whether a glyph is a .notdef?)

Reply:
If cid == 0 then it is the .notdef glyph so we shouldn't ever get 0 except for this one.

Reply:
@punchcutter `atoi()` returns 0 when its fed a string that doesn't contain a number, so if we can't verify here that we *should* be getting zero when we actually get zero we have a data sanitizing problem.

Reply:
We can hope the associated glyph name is either cid00000 or .notdef, but we can't guarantee it so we need some other way to verify that zero means zero.

Reply:
Probably best to switch to strtol/strtoul everywhere and check the `errno` value when it returns 0.

Reply:
@skef latest commit replaces strtol/strtoul/strtod with strtolCheck/strtoulCheck/strtodCheck functions